site/generator/publisher.py

```diff
@@ -0,0 +1,135 @@
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Union
+import dataclasses
+import io
+import os
+import re
+import shutil
+import sys
+
+from scss.compiler import compile_file
+import click
+
+from generator import models
+from generator import env
+
+
+class Publisher:
+    """Class for writing the AIP site out to disk."""
+
+    def __init__(self, output_dir, *, log_to_stdout=False):
+        self.output_dir = output_dir.rstrip(os.path.sep)
+        self.log_to_stdout = log_to_stdout
+        self.site = models.Site.load()
+
+    def log(self, message: str, *, err: bool = False):
+        """Conditionally log a message."""
+        dest = sys.stderr if err else sys.stdout
+        if self.log_to_stdout:
+            print(f'{message}', file=dest)
+
+    def publish_site(self):
```

Comment:
Assuming I'm reading correctly:

If publish is writing to disk, and app.py is serving from disk, I am more convinced we can keep the `assets` folder instead of `static` and reduce the migration costs.

Reply:
I may have engaged in some unintentional slight of hand here.

publish.py publishes _a static site_. In other words, it publishes something where you could point your browser at that folder (with no webserver) and it would work. This is roughly identical to the Jekyll output that the existing `serve.sh` hides from you by keeping it in a Docker image. In fact, I had to consult the Jekyll output to ensure I was doing the right thing.

By contrast, `app.py` is a development server that reads from the Markdown files in real-time and serves them. Like Jekyll, it is only meant as a dev server and there is no production server equivalent (publish.py generates a _static_ site). Unlike Jekyll, however, app.py does not work by generating a site on disk and then serving it; it just loads the files it needs into memory and does the conversation on the fly.